File: Sem3/Advanced Methods of Programming/MyInterpreter - Java Console Application/project/src/repository/Repository.java
package repository;


import domain.ProgramState;
import domain.adt.AdtException;
import domain.adt.MyList;
import domain.adt.MyListInterface;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;


public class Repository implements RepositoryInterface
{
    private String logFilePath;
    private List<ProgramState> programList;


    public Repository(String logFilePath) throws RepositoryException
    {
        programList = new ArrayList<>();
        this.logFilePath = logFilePath;
        try
        {
            PrintWriter logFile = new PrintWriter(new BufferedWriter(new FileWriter(logFilePath, false)));
            logFile.println("");
            logFile.close();
        }
        catch (IOException exception)
        {
            throw new RepositoryException("Cannot write to file." + logFilePath);
        }
    }


    @Override
    public void addProgram(ProgramState programState)
    {
        programList.add(programState);
    }


    @Override
    public void logProgramStateExec(ProgramState programState) throws RepositoryException
    {
        try
        {
            PrintWriter logFile = new PrintWriter(new BufferedWriter(new FileWriter(logFilePath, true)));
            logFile.println(programState.toString());
            logFile.close();
        }
        catch (IOException exception)
        {
            throw new RepositoryException("Cannot write to file." + logFilePath);
        }
    }


    @Override
    public ProgramState getCurrentProgram()
    {
        return programList.get(0);
    }


    @Override
    public List<ProgramState> getProgramList()
    {
        return programList;
    }


    @Override
    public void setProgramList(List<ProgramState> programList)
    {
        this.programList = programList;
    }
}

File: Sem3/Advanced Methods of Programming/MyInterpreter - CSharp Console Application/MyInterpreter CSharp/repository/Repository.cs
using System.IO;
using MyInterpreter_CSharp.domain;
using MyInterpreter_CSharp.domain.adt;

namespace MyInterpreter_CSharp.repository
{
    public class Repository : IRepository
    {
        private IMyList<ProgramState> _states;
        private readonly string _logFilePath;

        public Repository(string logFilePath)
        {
            _states = new MyList<ProgramState>();
            _logFilePath = logFilePath;
            File.WriteAllText(_logFilePath, "");
        }


        public void AddProgram(ProgramState programState)
        {
            _states.Add(programState);
        }

        public ProgramState GetCurrentProgram()
        {
            try
            {
                return _states.GetFirstElement();
            }
            catch (AdtException)
            {
                throw new RepositoryException("Empty repository.");
            }
        }

        public void LogProgramStateExec()
        {
            using (StreamWriter file = new StreamWriter(_logFilePath, true))
            {
                ProgramState programState = _states.GetFirstElement();
                file.WriteLine(programState);
            }
        }
    }       
 }